hand-detection/src/mediapipe/detector.ts

```diff
@@ -0,0 +1,151 @@
+/**
+ * @license
+ * Copyright 2021 Google LLC. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * =============================================================================
+ */
+import * as hands from '@mediapipe/hands';
+
+import {MEDIAPIPE_KEYPOINTS} from '../constants';
+import {HandDetector} from '../hand_detector';
+import {Keypoint} from '../shared/types';
+import {Hand, HandDetectorInput} from '../types';
+
+import {validateModelConfig} from './detector_utils';
+import {MediaPipeHandsMediaPipeEstimationConfig, MediaPipeHandsMediaPipeModelConfig} from './types';
+
+/**
+ * MediaPipe detector class.
+ */
+class MediaPipeHandsMediaPipeDetector implements HandDetector {
+  private readonly handsSolution: hands.Hands;
+
+  // This will be filled out by asynchronous calls to onResults. They will be
+  // stable after `await send` is called on the hands solution.
+  private width = 0;
+  private height = 0;
+  private hands: Hand[];
+
+  private selfieMode = false;
+
+  // Should not be called outside.
+  constructor(config: MediaPipeHandsMediaPipeModelConfig) {
+    this.handsSolution = new hands.Hands({
+      locateFile: (path, base) => {
+        if (config.solutionPath) {
+          const solutionPath = config.solutionPath.replace(/\/+$/, '');
+          return `${solutionPath}/${path}`;
+        }
+        return `${base}/${path}`;
+      }
+    });
+    this.handsSolution.setOptions({
+      selfieMode: this.selfieMode,
+      maxNumHands: config.maxHands,
+    });
+    this.handsSolution.onResults((results) => {
+      this.height = results.image.height;
+      this.width = results.image.width;
+      this.hands = [];
+      if (results.multiHandLandmarks !== null) {
+        const handednessList = results.multiHandedness;
+        const landmarksList = results.multiHandLandmarks;
+
+        for (let i = 0; i < handednessList.length; i++) {
+          this.hands.push({
+            keypoints: this.translateOutput(landmarksList[i]),
+            score: handednessList[i].score,
+            handedness: handednessList[i].label
+          });
+        }
+      }
+    });
+  }
+
+  private translateOutput(landmarks: hands.NormalizedLandmarkList): Keypoint[] {
+    return landmarks.map((landmark, i) => ({
+                           x: landmark.x * this.width,
+                           y: landmark.y * this.height,
+                           z: landmark.z,
```

Comment:
I see some inconsistency here. We usually scale Z landmarks similarly to x to preserve the same ratio in 3D. Here we clearly loose this scale information for z. Please check.